# This script will use edgeR to identify differentially accessible peaks based on read counts generated by diffbind

####Setup####

#set pvalue and fold change cutoffs
foldChange <- 0
pval <- 1e-4

setwd(dirname(rstudioapi::getActiveDocumentContext()$path))

library(edgeR)
library(DiffBind)
library(GenomicRanges)
library(rtracklayer)

#load regen object generated from diff bind, that has the read counts per peak
load("resources/untreated_ATAC_Counts.rds")


regenATAC <- dba.count(regenATAC, peaks=NULL, score=DBA_SCORE_READS)
counts <- dba.peakset(regenATAC, bRetrieve=TRUE, DataType=DBA_DATA_FRAME)

#drop peak coordinate columns
counts <- counts[,4:ncol(counts)]

#save raw counts file in csv format
write.csv(counts,file = "resources/untreated_ATAC_Counts.csv")

counts <- counts[,!(grepl("i",colnames(counts)))]

#fix column names
colnames(counts) <- gsub("X", "", colnames(counts))

#add row with gene IDs
counts$ID <- rownames(counts)

#bring in peak annotations
Annotations <- read.csv("resources/expanded_dovetail_SP_annot.csv", stringsAsFactors = F, row.names = 1)

peakLink <- read.table("resources/untreated_consensus_finalhits.txt", header = T, stringsAsFactors = F)
peakLink <- merge(peakLink, Annotations, by = "ID", all.x = T)
peaklink.og <- peakLink
peakLink <- peakLink[,c(1:5,19:23)]
colnames(peakLink) <- c("gene_ID","Chr_ID","Start", "End", "peak_id", colnames(peakLink)[6:10])

####Configure edgeR Analysis####

#define treatment groups
TR <- factor(c(rep("H0", 5), rep("F0", 5),
               rep("H3", 3), rep("F3", 3),
               rep("H8", 5), rep("F8", 5),
               rep("H12",5), rep("F12",5)))

#initiate DGElist object
analysisList <- DGEList(counts = counts[,1:(ncol(counts)-1)], group = TR, genes = counts$ID)

design <- model.matrix(~0+TR, data = analysisList$samples)

colnames(design) <- levels(analysisList$samples$group)

#exclude lowly expressed genes from the analysis 
#(at least three samples have at least ten counts per million)
keep <- rowSums(cpm(analysisList)>10) >= 3
analysisList <- analysisList[keep, , keep.lib.sizes=FALSE]

#calculate normalization factors for each sample
analysisList <- calcNormFactors(analysisList)

#calculate differential gene expression
analysisList <- estimateGLMRobustDisp(analysisList,design)

fit <- glmQLFit(analysisList,design, robust = T)

#set up the contrasts
my.contrasts <- makeContrasts(
  F0H0 = `F0` - `H0`,
  F3H3 = `F3` - `H3`,
  F8H8 = `F8` - `H8`,
  F3F0 = `F3` - `F0`,
  F8F0 = `F8` - `F0`,
  F8F3 = `F8` - `F3`,
  F12F0 = `F12` - `F0`,
  F12F8 = `F12` - `F8`,
  H3H0 = `H3` - `H0`,
  H8H0 = `H8` - `H0`,
  H8H3 = `H8` - `H3`,
  H12H0 = `H12` - `H0`,
  H12H8 = `H12` - `H8`,
  
  
  HRFR3c0 = (`H3` - `H0`) - (`F3` - `F0`),
  HRFR8c0 = (`H8` - `H0`) - (`F8` - `F0`),
  HRFR8c3 = (`H8` - `H3`) - (`F8` - `F3`),
  HRFR12c0 = (`H12` - `H0`) - (`F12` - `F0`),
  
  levels=design)

####Perform Pairwise Comparisons####

#function to generate the results objects for a named pairwise comparison

genResTables <- function(x,pv,fc) {
  #pull full results
  y <- glmQLFTest(fit, contrast = my.contrasts[,x])
  
  #extract results table
  z <- y$table
  z$ID <- rownames(z)
  z <- merge(z,peakLink, by.x = "ID", by.y = "peak_id", all.x = T)
  z$FDR <- p.adjust(z$PValue, method = "BH")
  
  #significance filters
  z.DG <- z[z$FDR <= pv,]
  z.DG.up <- z.DG[z.DG$logFC > fc,]
  z.DG.down <- z.DG[z.DG$logFC < -fc,]
  
  #return the four results objects
  return(list(z,z.DG,z.DG.up,z.DG.down))
}

#generate all results objects
#systematize object naming based on contrast names

for (i in 1:length(colnames(my.contrasts))) {
  
  resList <- genResTables(colnames(my.contrasts)[i],pval,foldChange)
  
  assign(colnames(my.contrasts)[i], resList[[1]])
  assign(paste0(colnames(my.contrasts)[i],".DG"), resList[[2]])
  assign(paste0(colnames(my.contrasts)[i],".DG.up"), resList[[3]])
  assign(paste0(colnames(my.contrasts)[i],".DG.down"), resList[[4]])
  
}

####Save the results####

#We also want to save the normalized counts (for plotting purposes)
normalizedCounts <- as.data.frame(cpm(analysisList, normalized.lib.sizes=T, log = T))

#save results dataframes, the DGE object, Annotations, and the normalized counts

#save results from all contrasts
saveThese <- ls(pattern = paste(colnames(my.contrasts),collapse = "|"))

# save gene lists in a csv (more git friendly format)
lapply(saveThese, function(x) write.csv(eval(parse(text=x)), file = paste0("Analysis_Output/ATAC/",x,".csv")))

#save contrasts table to help make sense of file names
write.csv(my.contrasts, file = "Analysis_Output/ATAC/contrasts.csv")

#some more useful things to save in Rdata format
saveThese <- c(saveThese,"analysisList","normalizedCounts","peakLink","regenATAC")

save(file = "Analysis_Output/ATAC/ATAC_DGE.RData", list = saveThese)


#export the peaks that changed significantly in response to injury (relative to 0hpa)

writeBed <- function(x,y) {
  z <- data.frame(chrom = x[,1], start = x[,2], end = x[,3], name = rownames(x), score = x[,4], strand = ".")
  write.table(z, file = y, quote = F, sep = "\t", row.names = F, col.names = F)
}

fullPeaks <- regenATAC$peaks[[1]]
fullPeaks$peak_id <- rownames(fullPeaks)

sigPeaks <- fullPeaks[fullPeaks$peak_id %in% c(H3H0.DG$ID,H8H0.DG$ID,H12H0.DG$ID,
                                               F3F0.DG$ID,F8F0.DG$ID,F12F0.DG$ID),]


writeBed(sigPeaks, "resources/regen_peakset.bed")
