# This script will use edgeR to identify differentially accessible peaks based on read counts generated by diffbind

####Setup####

#set pvalue and fold change cutoffs
foldChange <- 0
pval <- 1e-4

setwd(dirname(rstudioapi::getActiveDocumentContext()$path))

library(edgeR)
library(DiffBind)
library(GenomicRanges)
library(rtracklayer)

#load regen object generated from diff bind, that has the read counts per peak
load("resources/full_ATAC_Counts.rds")


regenATAC <- dba.count(regenATAC, peaks=NULL, score=DBA_SCORE_READS)
counts <- dba.peakset(regenATAC, bRetrieve=TRUE, DataType=DBA_DATA_FRAME)

#drop peak coordinate columns
counts <- counts[,4:ncol(counts)]
counts <- counts[,grepl("i",colnames(counts))]

#fix column names
colnames(counts) <- gsub("X", "", colnames(counts))

#add row with gene IDs
counts$ID <- rownames(counts)

#bring in peak annotations
Annotations <- read.csv("resources/expanded_dovetail_SP_annot.csv", stringsAsFactors = F, row.names = 1)

peakLink <- read.table("resources/full_consensus_finalhits.txt", header = T, stringsAsFactors = F)
peakLink <- merge(peakLink, Annotations, by = "ID", all.x = T)
peaklink.og <- peakLink
peakLink <- peakLink[,c(1:5,19:23)]
colnames(peakLink) <- c("gene_ID","Chr_ID","Start", "End", "peak_id", colnames(peakLink)[6:10])

####Configure edgeR Analysis####

#define treatment groups
TR <- factor(c(rep("Hi0", 4), rep("Fi0", 4),
               rep("Hi3", 4), rep("Fi3", 5),
               rep("Hi8", 5), rep("Fi8", 4),
               rep("Hi12",4), rep("Fi12", 5)))

#initiate DGElist object
analysisList <- DGEList(counts = counts[,1:(ncol(counts)-1)], group = TR, genes = counts$ID)

design <- model.matrix(~0+TR, data = analysisList$samples)

colnames(design) <- levels(analysisList$samples$group)

#exclude lowly expressed genes from the analysis 
#(at least three samples have at least ten counts per million)
keep <- rowSums(cpm(analysisList)>10) >= 3
analysisList <- analysisList[keep, , keep.lib.sizes=FALSE]

#calculate normalization factors for each sample
analysisList <- calcNormFactors(analysisList)

#calculate differential gene expression
analysisList <- estimateGLMRobustDisp(analysisList,design)

fit <- glmQLFit(analysisList,design, robust = T)

#set up the contrasts
my.contrasts <- makeContrasts(
  Fi3Fi0 = `Fi3` - `Fi0`,
  Hi3Hi0 = `Hi3` - `Hi0`,
  Fi8Fi0 = `Fi8` - `Fi0`,
  Hi8Hi0 = `Hi8` - `Hi0`,
  Fi12Fi0 = `Fi12` - `Fi0`,
  Hi12Hi0 = `Hi12` - `Hi0`,
  Fi8Fi3 = `Fi8` - `Fi3`,
  Hi8Hi3 = `Hi8` - `Hi3`,
  
  inhib8c0 = (`Hi8` - `Hi0`) - (`Fi8` - `Fi0`),
  inhib12c0 = (`Hi12` - `Hi0`) - (`Fi12` - `Fi0`),
  inhib3 = (`Hi3` - `Hi0`) - (`Fi3` - `Fi0`),
  levels=design)

####Perform Pairwise Comparisons####

#function to generate the results objects for a named pairwise comparison

genResTables <- function(x,pv,fc) {
  #pull full results
  y <- glmQLFTest(fit, contrast = my.contrasts[,x])
  
  #extract results table
  z <- y$table
  z$ID <- rownames(z)
  z <- merge(z,peakLink, by.x = "ID", by.y = "peak_id", all.x = T)
  z$FDR <- p.adjust(z$PValue, method = "BH")
  
  #significance filters
  z.DG <- z[z$FDR <= pv,]
  z.DG.up <- z.DG[z.DG$logFC > fc,]
  z.DG.down <- z.DG[z.DG$logFC < -fc,]
  
  #return the four results objects
  return(list(z,z.DG,z.DG.up,z.DG.down))
}

#generate all results objects
#systematize object naming based on contrast names

for (i in 1:length(colnames(my.contrasts))) {
  
  resList <- genResTables(colnames(my.contrasts)[i],pval,foldChange)
  
  assign(colnames(my.contrasts)[i], resList[[1]])
  assign(paste0(colnames(my.contrasts)[i],".DG"), resList[[2]])
  assign(paste0(colnames(my.contrasts)[i],".DG.up"), resList[[3]])
  assign(paste0(colnames(my.contrasts)[i],".DG.down"), resList[[4]])
  
}

####Save the results####

#We also want to save the normalized counts (for plotting purposes)
inormalizedCounts <- as.data.frame(cpm(analysisList, normalized.lib.sizes=T, log = T))

#save results dataframes, the DGE object, Annotations, and the normalized counts

#save results from all contrasts
saveThese <- ls(pattern = paste(colnames(my.contrasts),collapse = "|"))

# save gene lists in a csv (more git friendly format)
lapply(saveThese, function(x) write.csv(eval(parse(text=x)), file = paste0("Analysis_Output/ATAC/",x,".csv")))

#save contrasts table to help make sense of file names
write.csv(my.contrasts, file = "Analysis_Output/ATAC/icontrasts.csv")

ianalysisList <- analysisList

#some more useful things to save in Rdata format
saveThese <- c(saveThese,"ianalysisList","inormalizedCounts")

save(file = "Analysis_Output/ATAC/iATAC_DGE.RData", list = saveThese)



