#This script performs a DGE analysis on the read counts generated by RSEM

####Setup####

library(edgeR)
library(rstudioapi)
library(gsubfn)
library(ggfortify)

#decide on cutoffs

foldChange <- 0
pval <- 1e-3

#set the working directory to be the folder in which this script is located
setwd(dirname(getActiveDocumentContext()$path))

#load annotations
Annotations <- read.csv("resources/expanded_dovetail_SP_annot.csv", stringsAsFactors = F, row.names = 1)

####Configure edgeR Analysis####

#load read counts (genome)
counts <- read.table("resources/RNA.counts.matrix")

#rename sample names in columns and reorder
colnames(counts) <- c("0FR1", "0FR2", "0FR3",
                      "0HR1", "0HR2", "0HR3",
                      "12FR1","12FR2","12FR3",
                      "12HR1", "12HR2", "12HR3",
                      "3FR1", "3FR2", "3FR3",
                      "3HR1","3HR2","3HR3",
                      "8FR1","8FR2","8FR3",
                      "8HR1","8HR2","8HR3",
                      "0iFR1", "0iFR2", "0iFR3",
                      "0iHR1", "0iHR2", "0iHR3",
                      "12iFR1","12iFR2","12iFR3",
                      "12iHR1", "12iHR2", "12iHR3",
                      "8iFR1","8iFR2","8iFR3",
                      "8iHR1","8iHR2","8iHR3")

counts <- counts[,!grepl("i",colnames(counts))]

#add row with gene IDs
counts$ID <- rownames(counts)

#define treatment groups
TR <- factor(c(rep("F0", 3), rep("H0", 3),
               rep("F12", 3), rep("H12", 3),
               rep("F3", 3), rep("H3", 3),
               rep("F8", 3), rep("H8", 3)))

#initiate DGElist object
analysisList <- DGEList(counts = counts[,1:(ncol(counts)-1)], group = TR, genes = counts$ID)

#exclude lowly expressed genes from the analysis 
#(at least three samples have at least two counts per million)
keep <- rowSums(cpm(analysisList)>=2) >= 3
analysisList <- analysisList[keep, , keep.lib.sizes=F]

#calculate normalization factors for each sample
analysisList <- calcNormFactors(analysisList)

#define treatment groups
design <- model.matrix(~0+TR, data = analysisList$samples)

colnames(design) <- gsub("TR","",colnames(design))

#calculate differential gene expression
analysisList <- estimateGLMRobustDisp(analysisList,design)

fit <- glmQLFit(analysisList, design, robust = T)

#define pairwise comparisons to be used
my.contrasts <- makeContrasts(
  H0vF0 = `H0` - `F0`,
  H3vF3 = `H3` - `F3`,
  H8vF8 = `H8` - `F8`,
  H12vF12 = `H12` - `F12`,
  
  H3vH0 = `H3` - `H0`,
  H8vH0 = `H8` - `H0`,
  H12vH0 = `H12` - `H0`,
  F3vF0 = `F3` - `F0`,
  F8vF0 = `F8` - `F0`,
  F12vF0 = `F12` - `F0`,
  
  H8vH3 = `H8` - `H3`,
  F8vF3 = `F8` - `F3`,
  
  H12vH8 = `H12` - `H8`,
  F12vF8 = `F12` - `F8`,
  
  HR3vFR3 = (`H3` - `H0`) - (`F3` - `F0`),
  HR8vFR8c0 = (`H8` - `H0`) - (`F8` - `F0`),
  HR8vFR8c3 = (`H8` - `H3`) - (`F8` - `F3`),
  
  HR12vFR12c0 = (`H12` - `H0`) - (`F12` - `F0`),
  HR12vFR12c3 = (`H12` - `H3`) - (`F12` - `F3`),
  
  levels=design)

####Perform Pairwise Comparisons####

#function to generate the results objects for a named pairwise comparison

genResTables <- function(x,pv,fc) {
  #pull full results
  y <- glmQLFTest(fit, contrast = my.contrasts[,x])
  
  #extract results table
  z <- y$table
  z$ID <- rownames(z)
  z <- merge(z,Annotations, by = "ID", all.x = T)
  z$FDR <- p.adjust(z$PValue, method = "BH")
  
  #significance filters
  z.DG <- z[z$FDR <= pv,]
  z.DG.up <- z.DG[z.DG$logFC > fc,]
  z.DG.down <- z.DG[z.DG$logFC < -fc,]
  
  #return the four results objects
  return(list(z,z.DG,z.DG.up,z.DG.down))
}

#generate all results objects
#systematize object naming based on contrast names

for (i in 1:length(colnames(my.contrasts))) {
  
 resList <- genResTables(colnames(my.contrasts)[i],pval,foldChange)
 
 assign(colnames(my.contrasts)[i], resList[[1]])
 assign(paste0(colnames(my.contrasts)[i],".DG"), resList[[2]])
 assign(paste0(colnames(my.contrasts)[i],".DG.up"), resList[[3]])
 assign(paste0(colnames(my.contrasts)[i],".DG.down"), resList[[4]])
  
}

####Save the results####

#We also want to save the normalized counts (for plotting purposes)
normalizedCounts <- as.data.frame(cpm(analysisList, normalized.lib.sizes=T, log = T))

#save results dataframes, the DGE object, Annotations, and the normalized counts

#save results from all contrasts
saveThese <- ls(pattern = paste(colnames(my.contrasts),collapse = "|"))

# save gene lists in a csv (more git friendly format)
lapply(saveThese, function(x) write.csv(eval(parse(text=x)), file = paste0("Analysis_Output/RNA/",x,".csv")))

#save contrasts table to help make sense of file names
write.csv(my.contrasts, file = "Analysis_Output/RNA/contrasts.csv")

saveThese <- c(saveThese,"analysisList","normalizedCounts","Annotations")

save(file = "Analysis_Output/RNA/RNA_DGE.RData", list = saveThese)
