#This script performs a DGE analysis on the read counts generated by RSEM

####Setup####

library(edgeR)
library(rstudioapi)
library(gsubfn)
library(ggfortify)

#decide on cutoffs

foldChange <- 0
pval <- 5e-2

#set the working directory to be the folder in which this script is located
setwd(dirname(getActiveDocumentContext()$path))

#load annotations
Annotations <- read.csv("resources/expanded_dovetail_SP_annot.csv", stringsAsFactors = F, row.names = 1)

####Configure edgeR Analysis####

#load read counts (genome)
counts <- read.table("resources/wRNA.counts.matrix")

#rename sample names in columns and reorder
colnames(counts) <- c("0wFR1", "0wFR2", "0wFR3",
                      "0wHR1", "0wHR2", "0wHR3",
                      "16wFR1","16wFR2","16wFR3",
                      "16wHR1", "16wHR2", "16wHR3",
                      "2wFR1", "2wFR2", "2wFR3",
                      "2wHR1","2wHR2","2wHR3",
                      "4wFR1", "4wFR2", "4wFR3",
                      "4wHR1", "4wHR2", "4wHR3",
                      "8wFR1","8wFR2","8wFR3",
                      "8wHR1","8wHR2","8wHR3")

#add row with gene IDs
counts$ID <- rownames(counts)

#define treatment groups
TR <- factor(c(rep("F0", 3), rep("H0", 3),
               rep("F16", 3), rep("H16", 3),
               rep("F2", 3), rep("H2", 3),
               rep("F4", 3), rep("H4", 3),
               rep("F8", 3), rep("H8", 3)))

#initiate DGElist object
analysisList <- DGEList(counts = counts[,1:(ncol(counts)-1)], group = TR, genes = counts$ID)

#exclude lowly expressed genes from the analysis 
#(at least three samples have at least two counts per million)
keep <- rowSums(cpm(analysisList)>=2) >= 3
analysisList <- analysisList[keep, , keep.lib.sizes=F]

#calculate normalization factors for each sample
analysisList <- calcNormFactors(analysisList)

design <- model.matrix(~0+TR, data = analysisList$samples)

colnames(design) <- gsub("TR","",colnames(design))

#calculate differential gene expression
analysisList <- estimateGLMRobustDisp(analysisList,design)

fit <- glmQLFit(analysisList, design, robust = T)

#define pairwise comparisons to be used
my.contrasts <- makeContrasts(
  HRw16vFRw16c0 = (`H16` - `H0`) - (`F16` - `F0`),
  HRw8vFRw8c0 = (`H8` - `H0`) - (`F8` - `F0`),
  HRw4vFRw4c0 = (`H4` - `H0`) - (`F4` - `F0`),
  HRw2vFRw2c0 = (`H2` - `H0`) - (`F2` - `F0`),
  
  levels=design)

####Perform Pairwise Comparisons####

#function to generate the results objects for a named pairwise comparison

genResTables <- function(x,pv,fc) {
  #pull full results
  y <- glmQLFTest(fit, contrast = my.contrasts[,x])
  
  #extract results table
  z <- y$table
  z$ID <- rownames(z)
  z <- merge(z,Annotations, by = "ID", all.x = T)
  z$FDR <- p.adjust(z$PValue, method = "BH")
  
  #significance filters
  z.DG <- z[z$FDR <= pv,]
  z.DG.up <- z.DG[z.DG$logFC > fc,]
  z.DG.down <- z.DG[z.DG$logFC < -fc,]
  
  #return the four results objects
  return(list(z,z.DG,z.DG.up,z.DG.down))
}

#generate all results objects
#systematize object naming based on contrast names

for (i in 1:length(colnames(my.contrasts))) {
  
 resList <- genResTables(colnames(my.contrasts)[i],pval,foldChange)
 
 assign(colnames(my.contrasts)[i], resList[[1]])
 assign(paste0(colnames(my.contrasts)[i],".DG"), resList[[2]])
 assign(paste0(colnames(my.contrasts)[i],".DG.up"), resList[[3]])
 assign(paste0(colnames(my.contrasts)[i],".DG.down"), resList[[4]])
  
}

####Save the results####

#We also want to save the normalized counts (for plotting purposes)
wNormalizedCounts <- as.data.frame(cpm(analysisList, normalized.lib.sizes=T, log = T))

#save results dataframes, the DGE object, Annotations, and the normalized counts

#save results from all contrasts
saveThese <- ls(pattern = paste(colnames(my.contrasts),collapse = "|"))

# save gene lists in a csv (more git friendly format)
lapply(saveThese, function(x) write.csv(eval(parse(text=x)), file = paste0("Analysis_Output/RNA/",x,".csv")))

#save contrasts table to help make sense of file names
write.csv(my.contrasts, file = "Analysis_Output/RNA/wenger_contrasts.csv")

wAnalysisList <- analysisList

saveThese <- c(saveThese,"wAnalysisList","wNormalizedCounts")

save(file = "Analysis_Output/RNA/wRNA_DGE.RData", list = saveThese)
